File: java/base/src/main/vsdk/toolkit/gui/widget/WidgetDialog.java
package vsdk.toolkit.gui.widget;

import java.util.ArrayList;

import vsdk.toolkit.common.VSDK;
import vsdk.toolkit.common.logging.Logger;

/**
This class plays a role of internal node on an n-ary tree in the composite
design pattern.
*/
public class WidgetDialog extends WidgetElement {

    public static final int ORIENTATION_HORIZONTAL = 0x01;
    public static final int ORIENTATION_VERTICAL = 0x02;
    private String id;
    private String name;
    private int orientation;
    private ArrayList<WidgetElement> children;
    /// In the importing from file process, a dialog could be "incomplete",
    /// due to having a variable reference to a variable that is not loaded
    /// yet.  In such situation, a two pass processing is performed:
    /// On first pass does not create neither associate any variable, but store
    /// its incomplete references (names) on this ArrayList. On second pass
    /// the factory traverse this list in order to add current variables
    /// from context. Check WidgetPersistence.importAquynzaWidgetDialog method.
    private ArrayList<String> pendingVariableNames;
    private ArrayList<String> pendingCommandNames;
    private ArrayList<String> pendingDialogNames;
    private ArrayList<String> pendingDialogRefNames;
    private boolean collapsable;

    public WidgetDialog() {
        children = new ArrayList<WidgetElement>();
        pendingVariableNames = new ArrayList<String>();
        pendingCommandNames = new ArrayList<String>();
        pendingDialogNames = new ArrayList<String>();
        pendingDialogRefNames = new ArrayList<String>();
        orientation = ORIENTATION_VERTICAL;
        collapsable = false;
    }

    public boolean isCollapsable() {
        return collapsable;
    }

    public void setCollapsable(boolean collapsable) {
        this.collapsable = collapsable;
    }


    public ArrayList<String> getPendingCommandNames() {
        return pendingCommandNames;
    }

    public void setPendingCommandNames(ArrayList<String> pendingCommandNames) {
        this.pendingCommandNames = pendingCommandNames;
    }

    public ArrayList<WidgetElement> getWidgetElementList() {
        return children;
    }

    public void setWidgetElementList(ArrayList<WidgetElement> widgetElementList) {
        this.children = widgetElementList;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public double getOrientation() {
        return orientation;
    }

    public void setOrientation(int orientation) {
        this.orientation = orientation;
    }

    public ArrayList<String> getPendingVariableNames() {
        return pendingVariableNames;
    }

    public void setPendingVariableNames(ArrayList<String> pendingVariableNames) {
        this.pendingVariableNames = pendingVariableNames;
    }

    public ArrayList<String> getPendingDialogNames() {
        return pendingDialogNames;
    }

    public void setPendingDialogNames(ArrayList<String> pendingDialogNames) {
        this.pendingDialogNames = pendingDialogNames;
    }

    public ArrayList<WidgetElement> getChildren() {
        return children;
    }

    public void setChildren(ArrayList<WidgetElement> children) {
        this.children = children;
    }

    public ArrayList<String> getPendingDialogRefNames() {
        return pendingDialogRefNames;
    }

    public void setPendingDialogRefNames(ArrayList<String> pendingDialogRefNames) {
        this.pendingDialogRefNames = pendingDialogRefNames;
    }

    /**
     * Given a variableName, this method asks the context for a WidgetVariable
     * pointer in order to reference the given variable. If that variable
     * doesn't exist, an exception is thrown.
     */
    public void associateVariable(String variableName) {
        Logger.reportMessage(this, VSDK.FATAL_ERROR, "associateVariable",
                "Variable " + variableName + " not found!");
    }

    public void getPendingDialogNames(ArrayList<String> pendingDialogNames) {
        this.pendingDialogNames = pendingDialogNames;
    }

    @Override
    public String toString() {
        WidgetDialog dialog;
        String msg = "";

        msg = msg + "    DIALOG: " + this.getId() + "\n";
        for (int j = 0; j < children.size(); j++) {
            msg = msg + "    " + children.get(j).toString() + "\n";
        }
        for (int j = 0; j < pendingCommandNames.size(); j++) {
            msg = msg + "    Commandname: " + pendingCommandNames.get(j) + "\n";
        }

        for (int j = 0; j < pendingDialogRefNames.size(); j++) {
            msg = msg + "    DialogRefNames: " + pendingDialogRefNames.get(j) + "\n";
        }

        for (int j = 0; j < pendingVariableNames.size(); j++) {
            msg = msg + "    VariableNames: " + pendingVariableNames.get(j) + "\n";
        }
        return msg;
    }
}

File: cpp/base/src/main/vsdk/toolkit/gui/widget/WidgetDialog.cpp
#include "java/util/ArrayList.txx"
#include "vsdk/toolkit/common/logging/Logger.h"
#include "vsdk/toolkit/gui/widget/WidgetDialog.h"

WidgetDialog::WidgetDialog()
    : id(""), name(""), orientation(ORIENTATION_VERTICAL), collapsable(false)
{
}

bool WidgetDialog::isCollapsable() const
{
    return collapsable;
}

void WidgetDialog::setCollapsable(bool collapsable)
{
    this->collapsable = collapsable;
}

java::ArrayList<java::String>& WidgetDialog::getPendingCommandNames()
{
    return pendingCommandNames;
}

void WidgetDialog::setPendingCommandNames(
    const java::ArrayList<java::String>& pendingCommandNames)
{
    this->pendingCommandNames = pendingCommandNames;
}

java::ArrayList<WidgetElement*>& WidgetDialog::getWidgetElementList()
{
    return widgetElementList;
}

void WidgetDialog::setWidgetElementList(
    const java::ArrayList<WidgetElement*>& widgetElementList)
{
    this->widgetElementList = widgetElementList;
}

const java::String& WidgetDialog::getId() const
{
    return id;
}

void WidgetDialog::setId(const java::String& id)
{
    this->id = id;
}

const java::String& WidgetDialog::getName() const
{
    return name;
}

void WidgetDialog::setName(const java::String& name)
{
    this->name = name;
}

double WidgetDialog::getOrientation() const
{
    return orientation;
}

void WidgetDialog::setOrientation(int orientation)
{
    this->orientation = orientation;
}

java::ArrayList<java::String>& WidgetDialog::getPendingVariableNames()
{
    return pendingVariableNames;
}

void WidgetDialog::setPendingVariableNames(
    const java::ArrayList<java::String>& pendingVariableNames)
{
    this->pendingVariableNames = pendingVariableNames;
}

java::ArrayList<java::String>& WidgetDialog::getPendingDialogNames()
{
    return pendingDialogNames;
}

void WidgetDialog::setPendingDialogNames(
    const java::ArrayList<java::String>& pendingDialogNames)
{
    this->pendingDialogNames = pendingDialogNames;
}

java::ArrayList<WidgetElement*>& WidgetDialog::getChildren()
{
    return widgetElementList;
}

void WidgetDialog::setChildren(
    const java::ArrayList<WidgetElement*>& widgetElementList)
{
    this->widgetElementList = widgetElementList;
}

java::ArrayList<java::String>& WidgetDialog::getPendingDialogRefNames()
{
    return pendingDialogRefNames;
}

void WidgetDialog::setPendingDialogRefNames(
    const java::ArrayList<java::String>& pendingDialogRefNames)
{
    this->pendingDialogRefNames = pendingDialogRefNames;
}

void WidgetDialog::associateVariable(const java::String& variableName)
{
    Logger::reportMessage("WidgetDialog", Logger::FATAL_ERROR,
        "associateVariable",
        java::String("Variable ") + variableName + " not found!");
}

java::String WidgetDialog::toString() const
{
    java::String msg = "";

    msg = msg + "    DIALOG: " + getId() + "\n";
    for ( int j = 0; j < widgetElementList.size(); j++ ) {
        msg = msg + "    " + widgetElementList.get(j)->toString() + "\n";
    }
    for ( int j = 0; j < pendingCommandNames.size(); j++ ) {
        msg = msg + "    Commandname: " + pendingCommandNames.get(j) + "\n";
    }

    for ( int j = 0; j < pendingDialogRefNames.size(); j++ ) {
        msg = msg + "    DialogRefNames: " + pendingDialogRefNames.get(j) + "\n";
    }

    for ( int j = 0; j < pendingVariableNames.size(); j++ ) {
        msg = msg + "    VariableNames: " + pendingVariableNames.get(j) + "\n";
    }
    return msg;
}
